baseplate/metrics.py

```diff
@@ -255,18 +276,73 @@ def increment(self, delta=1, sample_rate=1.0):
         :param float delta: The amount to increase the counter by.
         :param float sample_rate: What rate this counter is sampled at. [0-1].
 
+        """
+        self.send(delta, sample_rate)
+
+    def decrement(self, delta=1, sample_rate=1.0):
+        """Decrement the counter.
+
+        This is equivalent to :py:meth:`increment` with delta negated.
+
+        """
+        self.increment(delta=-delta, sample_rate=sample_rate)
+
+    def send(self, delta, sample_rate):
+        """Send the counter to the backend.
+
+        :param float delta: The amount to increase the counter by.
+        :param float sample_rate: What rate this counter is sampled at. [0-1].
+
         """
         parts = [
             self.name + (":{:g}".format(delta).encode()),
             b"c",
         ]
 
-        if sample_rate and sample_rate != 1.0:
+        if sample_rate != 1.0:
             parts.append("@{:g}".format(sample_rate).encode())
 
         serialized = b"|".join(parts)
         self.transport.send(serialized)
 
+
+class BatchCounter(Counter):
+    """Counter implementation that batches multiple increment calls.
+
+    A new entry in the :term:`packets` entry is created for each sample rate.
+    For example, if a counter is incremented multiple times with a sample
+    rate of 1.0, there will be one entry in :term:`packets`. If that counter
+    is implemented again with a sample rate of 0.5, there will be two entries
+    in :term:`packets`. Each packet has an associated delta value.
+
+    Example usage::
+        counter = BatchCounter(transport, "counter_name")
+        counter.increment()
+        do_something_else()
+        counter.increment()
+        counter.send()
+
+    The above example results in one packet indicating an increase of two
+    should be applied to "counter_name".
+    """
+    def __init__(self, transport, name):
+        self.transport = transport
+        self.name = name
+        self.packets = {}
+
+    def increment(self, delta=1, sample_rate=1.0):
+        """Increment the counter.
+
+        :param float delta: The amount to increase the counter by.
+        :param float sample_rate: What rate this counter is sampled at. [0-1].
+
+        """
+        stale_delta = self.packets.get(sample_rate)
```

Comment:
a `collections.defaultdict(int)` would make this just `self.packets[sample_rate] += delta`